osquery/events/linux/audit.cpp

```diff
@@ -0,0 +1,286 @@
+      } else {
+        // Add this rule data to the publisher's list of transient rules.
+        // These will be removed during tear down or re-configure.
+      }
+    }
+  }
+
+  // The audit library provides an API to send a netlink request that fills in
+  // a netlink reply with audit rules. As such, this process will maintain a
+  // single open handle and reply to audit-metadata tables with the buffered
+  // content from the publisher.
+  if (audit_request_rules_list_data(handle_) <= 0) {
+    // Could not request audit rules.
+  }
+}
+
+void AuditEventPublisher::tearDown() {
+  if (handle_ > 0) {
+    audit_close(handle_);
+  }
+}
+
+inline void handleConfigChange(const struct audit_reply& reply) {
+  // Another daemon may have taken control.
+}
+
+inline bool handleReply(const struct audit_reply& reply,
+                        AuditEventContextRef& ec) {
+  // Build an event context around this reply.
+  ec->type = reply.type;
+
+  // Tokenize the message.
+  auto message = std::string(reply.message, reply.len);
+  std::vector<std::string> fields;
+  boost::split(fields, message, boost::is_any_of(" ="));
+  std::string().swap(message);
+
+  // Iterate over each field=value pair.
+  auto field_it = fields.begin();
+  ec->preamble = std::move(*field_it);
+  for (++field_it; field_it != fields.end(); field_it++) {
+    const auto& key = *field_it;
+    if (++field_it == fields.end()) {
+      // A malformed message will have had an odd number of fields=value
+      // pairs, discard the event.
+      return false;
+    }
+    // Assign the key/value pair.
+    ec->fields[key] = std::move(*field_it);
+  }
+
+  // There is a special field for syscalls.
+  if (ec->fields.count("syscall") == 1) {
+    const auto& syscall_string = ec->fields.at("syscall").c_str();
+    char* end = nullptr;
+    long long int syscall = strtoll(syscall_string, &end, 10);
+    if (end == nullptr || end == syscall_string || *end != '\0' ||
+        ((syscall == LLONG_MIN || syscall == LLONG_MAX) && errno == ERANGE)) {
+      syscall = 0;
+    }
+    ec->syscall = syscall;
+  }
+
+  return true;
+}
+
+void AuditEventPublisher::handleListRules() {
+  // Store the rules response.
+}
+
+Status AuditEventPublisher::run() {
+  if (count_ == 0 || count_++ % 10 == 0) {
+    // Request an update to the audit status.
+    // This will also fill in the status on first run.
+    audit_request_status(handle_);
+  }
+
+  // Reset the reply data.
+  int result = 0;
+  bool handle_reply = false;
+  while (true) {
+    handle_reply = false;
+
+    // Request a reply in a non-blocking mode.
+    // This allows the publisher's run loop to periodically request an audit
+    // status update. These updates can check for other processes attempting to
+    // gain control over the audit sink.
+    // This non-blocking also allows faster receipt of multi-message events.
+    result = audit_get_reply(handle_, &reply_, GET_REPLY_NONBLOCKING, 0);
```

Comment:
It's not clear who owns the member of the `reply` structure's member raw pointers? Valgrind plus a python script launching 1000 processes does not show any direct-loss of member. A review of Ubuntu's `auditd` and patches show no obvious freeing. Either way, more investigation is needed.